server/src/main/java/org/apache/druid/segment/realtime/appenderator/SinkQuerySegmentWalker.java

```diff
@@ -182,7 +182,7 @@ public <T> QueryRunner<T> getQueryRunnerForSegments(final Query<T> query, final
     final DataSourceAnalysis analysis = dataSourceFromQuery.getAnalysis();
 
     // Sanity check: make sure the query is based on the table we're meant to handle.
-    if (!analysis.getBaseTableDataSource().filter(ds -> dataSource.equals(ds.getName())).isPresent()) {
+    if (!analysis.getBaseTableDataSource().getName().equals(dataSource)) {
```

Comment:
```suggestion
    if (!analysis.hasBaseTableDataSource() || !analysis.getBaseTableDataSource().getName().equals(dataSource)) {
```

Reply:
I feel like that won't be much different:
I think:
* earlier the return type for `getBaseTableDataSource` was an `Optional` ; if it returned `empty` - it returned the below error; the new `getBaseTableDataSource` is not an `Optional` and has a check inside it; this leaves that case to that.
* if there were no match to the filter it become empty that way and returned the next error - now this become the equals check